server/src/main/java/org/candlepin/controller/CandlepinOwnerProductShareManager.java

```diff
@@ -0,0 +1,200 @@
+/**
+ * Copyright (c) 2009 - 2017 Red Hat, Inc.
+ *
+ * This software is licensed to you under the GNU General Public License,
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.controller;
+
+import org.candlepin.model.Owner;
+import org.candlepin.model.OwnerProductCurator;
+import org.candlepin.model.OwnerProductShare;
+import org.candlepin.model.OwnerProductShareCurator;
+import org.candlepin.model.Product;
+
+import com.google.inject.Inject;
+
+import org.apache.commons.collections.CollectionUtils;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Manages Shared product operations.
+ */
+public class CandlepinOwnerProductShareManager implements OwnerProductShareManager {
+
+    private OwnerProductCurator ownerProductCurator;
+    private OwnerProductShareCurator shareCurator;
+
+    @Inject
+    public CandlepinOwnerProductShareManager(OwnerProductCurator ownerProductCurator,
+        OwnerProductShareCurator shareCurator) {
+        this.ownerProductCurator = ownerProductCurator;
+        this.shareCurator = shareCurator;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Product resolveProductById(Owner recipientOwner, String productId, boolean includeShares) {
+        List<Product> products = resolveProductsByIds(recipientOwner,
+            Collections.singleton(productId), includeShares);
+        if (CollectionUtils.isNotEmpty(products)) {
+            return products.get(0);
+        }
+        return null;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public List<Product> resolveProductsByIds(Owner recipientOwner, Collection<String> productIds,
+        boolean includeShares) {
+        Map<String, Product> result = new HashMap<String, Product>();
+        List<Product> existingProducts;
+        Set<String> missingProducts = new HashSet<String>();
+
+        if (productIds != null && !productIds.isEmpty()) {
+            // add all requested now so we can remove the found ones later.
+            missingProducts.addAll(productIds);
+            existingProducts = ownerProductCurator.getProductsByIds(recipientOwner, productIds).list();
+        }
+        else {
+            existingProducts = ownerProductCurator.getProductsByOwner(recipientOwner).list();
```

Comment:
Is this meant to handle the case where a caller wants to resolve *all* of an owner products?  The semantics here seem odd.  Sending in an empty or null list of productIds seems like it should be an error (or at the very least a no-op) instead of an expected/accepted way of doing things.